easylogging: Remove sticky std::hex configuration and add showbase and boolalpha flags to logs

morantr opened the pull request:
Add to LoggingFlag on easylogging 3 new flags:
ForceDecBase - Set the log to print integer on decimal format with every log-entry.
ShowBase - Set the log to print the base of integer value according to iostream base flag.
BoolAlpha - Set the log to print the booleans as "true" or "false" instead of "1" or "0".

Set these configurations by default on all processes.
Now  the log will look like this:

LOG(DEBUG) << std::hex << "num_hex=" << 32 << ", bool=" << bool(false);
LOG(DEBUG) << "num_dec=" << 32 << ", bool=" << bool(true);

output:
```
DEBUG 16:33:21:731 <Some file> --> num_hex=0x20, bool=false
DEBUG 16:33:21:731 <Some file> --> num_dec=32, bool=true
```

#879

tomereli reviewed (commented):
I don't think this should be merged. This has nothing to do with easylogging - this is how the standard library works and I don't think we should change it.
But I'm open to other opinions - @arnout @rmelotte @vitalybu

morantr commented:
> I don't think this should be merged. This has nothing to do with easylogging - this is how the standard library works and I don't think we should change it.
> But I'm open to other opinions - @arnout @rmelotte @vitalybu

It has nothing with easylogging, but as I can see it, changing it on easylogging is the only way.
Right now it is very hard to debug things. For example on optimal path print of rssi measurement looks like this:
"rssi=ffffffc4"

So unless you want to manually add std::dec after every print, this is the way to go.

vitalybu reviewed (changes requested):
I agree with @tomereli. 
We should not modify external libraries source code.
If you think it's a useful change, you can try adding into the easylogging repo.

tomereli commented:
> > I don't think this should be merged. This has nothing to do with easylogging - this is how the standard library works and I don't think we should change it.
> > But I'm open to other opinions - @arnout @rmelotte @vitalybu
> 
> It has nothing with easylogging, but as I can see it, changing it on easylogging is the only way.
> Right now it is very hard to debug things. For example on optimal path print of rssi measurement looks like this:
> "rssi=ffffffc4"
> 
> So unless you want to manually add std::dec after every print, this is the way to go.

You can add `std::dec` where you think it is important for debugging that the print will be decimal, it certainly isn't the case in all cases. The real solution is to add std::dec in the places where we switched to std::hex and forgot to reset it.

morantr commented:
> > > I don't think this should be merged. This has nothing to do with easylogging - this is how the standard library works and I don't think we should change it.
> > > But I'm open to other opinions - @arnout @rmelotte @vitalybu
> > 
> > 
> > It has nothing with easylogging, but as I can see it, changing it on easylogging is the only way.
> > Right now it is very hard to debug things. For example on optimal path print of rssi measurement looks like this:
> > "rssi=ffffffc4"
> > So unless you want to manually add std::dec after every print, this is the way to go.
> 
> You can add `std::dec` where you think it is important for debugging that the print will be decimal, it certainly isn't the case in all cases. The real solution is to add std::dec in the places where we switched to std::hex and forgot to reset it.

It will work on theory, but developers will forget to do add it after switching to hex, and the log will be impossible to read. Also, it will be very hard to map all the places you need decimal value.
Is it possible to make clang-format to add `<< std::dec` at the end of a stream that contains `std::hex`?

tomereli commented:
@arnout please share your opinion

tomereli commented:
> > > > I don't think this should be merged. This has nothing to do with easylogging - this is how the standard library works and I don't think we should change it.
> > > > But I'm open to other opinions - @arnout @rmelotte @vitalybu
> > > 
> > > 
> > > It has nothing with easylogging, but as I can see it, changing it on easylogging is the only way.
> > > Right now it is very hard to debug things. For example on optimal path print of rssi measurement looks like this:
> > > "rssi=ffffffc4"
> > > So unless you want to manually add std::dec after every print, this is the way to go.
> > 
> > 
> > You can add `std::dec` where you think it is important for debugging that the print will be decimal, it certainly isn't the case in all cases. The real solution is to add std::dec in the places where we switched to std::hex and forgot to reset it.
> 
> It will work on theory, but developers will forget to do add it after switching to hex, and the log will be impossible to read. Also, it will be very hard to map all the places you need decimal value.
> Is it possible to make clang-format to add `<< std::dec` at the end of a stream that contains `std::hex`?

It's not the issue where to put it, it's if to put this at all.
This will break cases where user wants to dump multiple LOGs with std::hex for example:
```
LOG(DEBUG) << "start dump in hex" << std::hex ;
...
LOG(DEBUG) << "val=" << val;
...
LOG(DEBUG) << "val2=" << val2;
...
LOG(DEBUG) << "End dump in hex" << std::dec;
```

arnout commented:
> It will work on theory, but developers will forget to do add it after switching to hex
As mentioned before, this is standard library behaviour:
```
#include <iostream>

using namespace std;
int main ()
{
    cout << "hex: " << hex << 0x40 << " still hex: " << 100 << endl;
    cout << "still hex: " << 100 << endl;
    return 0;
}
```

That said, I agree that the standard C++ behaviour is extremely annoying. So I kind of agree with this change.  Only, I think doing it in the LOG macro is not appropriate. It is IMO more appropriate to do it in `MessageBuilder::initialize()` and to explicitly set all flags with `m_logger->stream().flags(...)`. We could also add some other useful flags like showbase.

tomereli commented:
> > It will work on theory, but developers will forget to do add it after switching to hex
> > As mentioned before, this is standard library behaviour:
> 
> ```
> #include <iostream>
> 
> using namespace std;
> int main ()
> {
>     cout << "hex: " << hex << 0x40 << " still hex: " << 100 << endl;
>     cout << "still hex: " << 100 << endl;
>     return 0;
> }
> ```
> 
> That said, I agree that the standard C++ behaviour is extremely annoying. So I kind of agree with this change. Only, I think doing it in the LOG macro is not appropriate. It is IMO more appropriate to do it in `MessageBuilder::initialize()` and to explicitly set all flags with `m_logger->stream().flags(...)`. We could also add some other useful flags like showbase.

I'm OK with this solution. We should not block changes in easylogging since we don't know if they will be accepted, so I suggest to implement this and then upstream all our changes in a single PR. @morantr I think it is OK if we just create a task for upstreaming the easylogging changes for now, and mark it as "good first issue" :)

ghost commented:
I still don't really get why people think working with `operator<<` and ostreams is so much better than using C streams and `fprintf`. At least, then you always know what you're going to get ...

dammit! I'm sounding like a grumpy old man again! And I swear I'm only 31 years old ...

morantr committed:
easylogging: Add new flags configuration

Add to LoggingFlag on easylogging 3 new flags:
ForceDecBase - Set the log to print integer on decimal format with every log-entry.
ShowBase - Set the log to print the base of integer value according to iostream base flag.
BoolAlpha - Set the log to print the booleans as "true" or "false" instead of "1" or "0".

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
easylogging: initialize new flags configuration if set

On the previous commit, three new configuration flags have been defined.
The new configurations are changing the visual presentation of integers on the logs.
Set these configuration if its flags are set.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: Configure easylogging with new configuration flags

Set ForceDecBase, ShowBase and BoolAlpha flags to easylogging configuration.
These configurations have been added to easylogging on the previous commit and in this commit,
they are set on beerocks_logging.cpp which set the configuration on all beerocks processes.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

tomereli commented:
Wait, so now this means I can never print in hex?

vitalybu commented on framework/external/easylogging/easylogging++.h:
```diff
@@ -725,8 +725,14 @@ enum class LoggingFlag : base::type::EnumType {
   /// @brief Preserves time format and does not convert it to sec, hour etc (performance tracking only)
   FixedTimeFormat = 16384,
-  // @brief Ignore SIGINT or crash
+  //  @brief Ignore SIGINT or crash
```
What's up with the extra space?

morantr replied:
All other similar lines were with like this, so I aligned to the rest of the lines.
Edit:
I now realize that I was missing slash. All other comments were with 3 of them, and I added only two. Fixed.

morantr commented:
> What will happen now if I do `LOG(DEBUG) <<std::hex<< x;`? Will it still print X in decimal instead of hex?

It will print in hex notation until the end of the log print call.
The difference is that now we don't need to change it back to `std::dec`. It will be done automatically before every new print.
Please see the PR description. I have updated it, and it explains it.

morantr committed:
tests: Update grep printing "test_higher_layer_data_payload_trigger"

Due to the change in of hex notation printing that now adds "0x" prefix to integers in hex format
in the previous commit, update few grep commands on integers which are being printed in hex format
to hex notation.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

vitalybu commented:
Overall I'm OK with this change, but as we introduce more and more changes into this library, maybe we should move to CMake's [ExternalProject](https://cmake.org/cmake/help/v3.8/module/ExternalProject.html) + our own patches.
